Source/Core/InputCommon/ControllerInterface/CemuHookUDPServer/CemuHookUDPServerProto.h

```diff
@@ -0,0 +1,262 @@
+  u32 hid_packet_counter;
+  u8 button_states1;
+  u8 button_states2;
+  u8 button_ps;
+  u8 button_touch;
+  u8 left_stick_x;
+  u8 left_stick_y_inverted;
+  u8 right_stick_x;
+  u8 right_stick_y_inverted;
+  u8 button_dpad_left_analog;
+  u8 button_dpad_down_analog;
+  u8 button_dpad_right_analog;
+  u8 button_dpad_up_analog;
+  u8 button_square_analog;
+  u8 button_cross_analog;
+  u8 button_circle_analog;
+  u8 button_triangle_analog;
+  u8 button_r1_analog;
+  u8 button_l1_analog;
+  u8 trigger_r2;
+  u8 trigger_l2;
+  Touch touch1;
+  Touch touch2;
+  u64 timestamp_us;
+  float accelerometer_x_g;
+  float accelerometer_y_g;
+  float accelerometer_z_inverted_g;
+  float gyro_pitch_deg_s;
+  float gyro_yaw_deg_s;
+  float gyro_roll_deg_s;
+};
+
+struct FromServer
+{
+  union
+  {
+    struct
+    {
+      MessageHeader header;
+      u32 message_type;
+    };
+    MessageType::VersionResponse version_response;
+    MessageType::PortInfo port_info;
+    MessageType::PadDataResponse pad_data_response;
+  };
+};
+
+struct FromClient
+{
+  union
+  {
+    struct
+    {
+      MessageHeader header;
+      u32 message_type;
+    };
+    MessageType::VersionRequest version_request;
+    MessageType::ListPorts list_ports;
+    MessageType::PadDataRequest pad_data_request;
+  };
+};
+}  // namespace MessageType
+
+static inline u32 CRC32(const void* buffer, unsigned length)
+{
+  return crc32(crc32(0L, Z_NULL, 0), (const Bytef*)buffer, length);
+}
+
+template <typename MsgType>
+struct Message : public MsgType
+{
+  Message() {}
+
+  Message(u32 source_uid)
+  {
+    memset(this, 0, sizeof *this);
+    static_assert(sizeof(char) == sizeof(u8));
+    strncpy((char*)MsgType::header.source, MsgType::FROM, sizeof MsgType::header.source);
+    MsgType::header.protocol_version = CEMUHOOK_PROTOCOL_VERSION;
+    MsgType::header.message_length = sizeof *this - sizeof MsgType::header;
+    MsgType::header.source_uid = source_uid;
+    MsgType::message_type = MsgType::TYPE;
+  }
+
+  void Finish() { MsgType::header.crc32 = CRC32(this, sizeof *this); }
+
+  template <class ToMsgType>
+  ToMsgType* CheckAndCastTo()
+  {
+    u32 crc32 = MsgType::header.crc32;
+    // zero out the crc32 in the packet once we got it since that's whats needed for calculation
+    MsgType::header.crc32 = 0;
+    if (crc32 != CRC32(this, sizeof(ToMsgType)))
+    {
+      NOTICE_LOG(SERIALINTERFACE,
+                 "CemuHookUDPServer Received message with bad CRC: got %u, expected %u", crc32,
+                 CRC32(this, sizeof(ToMsgType)));
+      return nullptr;
+    }
+    if (MsgType::header.protocol_version > CEMUHOOK_PROTOCOL_VERSION)
+      return nullptr;
+    if (memcmp(MsgType::header.source, ToMsgType::FROM, sizeof MsgType::header.source))
+      return nullptr;
+    if (MsgType::message_type != ToMsgType::TYPE)
+      return nullptr;
+    if (MsgType::header.message_length + sizeof MsgType::header > sizeof(ToMsgType))
+      return nullptr;
+    return (ToMsgType*)this;
```

Comment:
This is undefined behavior regardless of the checks that come before it. Please don't do this and either memcpy the data into the destination type, and return it by value, or find another way to structure the data.

Reply:
I changed the return type to std::optional<ToMsgType> and went with your alternative of memcpy'ing the data into the destination type and returning it. I guess this copies the returned value twice?